### PR TITLE
fix: historical Dataset Batch Processing DB Update

### LINKED ISSUE #169: Historical Dataset Batch Processing DB Update Bug
It seems we are not adding the new instances of `GTFSDataset` to the database properly. 
For `mdb-1` for example, here are the logs for November 9th:

Image

It mentions that the hash is NULL which should not happen since we have [multiple records for `mdb-1`](https://console.cloud.google.com/storage/browser/mobility-datasets/mdb-1?project=mobility-feeds-dev&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false) in the `mobility-datasets` bucket.

**Proposed solution:**
- [ ] Remove GTFSDataset reference from the populate script
- [ ] Introduce new functions : one responsible for extracting the bounding box from a dataset_id
- [ ] Change the existing functions to allow null dataset_id and dataset_hash
- [ ] Close session on batch_datasets function

### PR DESCRIPTION
**Pull request by davidgamez**
**Summary:**

Closes #169 
Part of #105

Sorry if I included too many changes at once; it was hard to have them independently.
Summary
- [x] Moved batch functions inside functions-python. The reasoning is to have a single place for all functions to reuse common patterns for build, testing, formatting, and sharing packages.
- [x] Adding a DB exclusively for testing. As part of each function testing, now it's possible to have more control over what the DB contains, making the tests less likely to fail due to local changes in the DB.
- [x] Created a _database_ utils file for creating DB session and shared across functions. Introduced the concept of reusing the DB connection. This helps to minimize the number of connections as the connections can be reused between different executions of the same function when GCP reuses the same container. The connection should be closed when GCP discards the container, but I haven't tried it yet...
- [x] Reduced the number of GH parameters by moving function-related parameters to JSON files and setting multiple parameters with default values.
- [x] Create a Python class for  the datastore entities
- [x] Reduce the number of HTTP retries. This was causing some feeds to be infinite get processed as the retries exceeded the pub/sub retry timeout.
- [x]  Create a DB URL per environment GCP secret to inject in the Python functions.
- [x] Batch Functions in dev and QA will only process 10 feeds to reduce cost
- [x] Fix: lint script was not failing the build
- [x] Fix: test  script was not failing the build
- [x] Feeds exposed by legacy SSL versions can be downloaded

**Expected behavior:** 

- [x] Feeds are not executed infinite
- [x] Datasets are not published when the previous dataset hash is the same
- [x] Remove the previous dependency on the existence of the dataset from the populate script. 

**Testing tips:**

- Select a feed and change the latest dataset hash field to any string
- On dev, go to google scheduler console 
- Select the `dataset-batch-job-dev` job 
- Click _Resume_
- Click _Force Run_
- Wait for all datasets to be process
- Check _gtfsdataset_ table and verify that a new dataset is created with a valid hash

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `./scripts/api-tests.sh` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: functions-python/batch_datasets/src/main.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            Gtfsdataset.id.label("dataset_id"),
+            Gtfsdataset.hash.label("dataset_hash"),
+        )
+        .select_from(Gtfsfeed)
+        .outerjoin(Gtfsdataset, (Gtfsdataset.feed_id == Gtfsfeed.id))
+        .filter(Gtfsfeed.status == "active", Gtfsfeed.authentication_type == "0")
+        .filter(or_(Gtfsdataset.id.is_(None), Gtfsdataset.latest.is_(True)))
+    )
+    # Limit the query to 10 feeds (or FEEDS_LIMIT param) for testing purposes and lower environments
+    if os.getenv("ENVIRONMENT", "").lower() in ("dev", "test", "qa"):
```
Comment (cka-y) [inline functions-python/batch_datasets/src/main.py:80]: Nice addition

File: functions-python/dataset_service/main.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        else:
+            return []
+
+    def _execution_to_entity(self, execution: BatchExecution) -> datastore.Entity:
+        key = self.client.key(batch_execution_collection, execution.execution_id)
+        entity = datastore.Entity(key=key)
+
+        entity.update(asdict(execution))
+
+        return entity
```
Comment (cka-y) [inline functions-python/dataset_service/main.py:136]: suggestion: some documentation for both `BatchExecutionService` and `DatasetTraceService` might be beneficial for better understanding

File: functions-python/batch_process_dataset/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            }            
+        }
+    }
+```
+
+# Function configuration
+The function is configured using the following environment variables:
+- `DATASETS_BUCKET_NANE`: The name of the bucket where the datasets are stored.
+- `FEEDS_DATABASE_URL`: The URL of the feeds database.
+- `MAXIMUM_EXECUTIONS`: [Optional] The maximum number of executions per datasets. This controls the number of times a dataset can be processed per execution id. By default, is 2.
```
[the remark this answers is not in the document]
Comment (davidgamez) [reply functions-python/batch_process_dataset/README.md:32]: As we removed the retry at the HTTP request level, I added it here at the function level. I'm happy to take it out if we agree that it is not necessary.

[the remark this answers is not in the document]
Comment (davidgamez) [reply functions-python/batch_process_dataset/README.md]: Done, set maximum to 1.

**Review by cka-y — APPROVED** [review]
LGTM!

**Pull request merged**